Tests/NIOTests/SocketChannelTest.swift

```diff
@@ -261,4 +261,27 @@ public class SocketChannelTest : XCTestCase {
         try serverChannel.close().wait()
     }
 
+
+    public func testWithConnectedSocket() throws {
+        let group = MultiThreadedEventLoopGroup(numThreads: 1)
+        defer { XCTAssertNoThrow(try group.syncShutdownGracefully()) }
+
+        let serverChannel = try ServerBootstrap(group: group).bind(host: "127.0.0.1", port: 0).wait()
+
+        let clientSock = try Socket(protocolFamily: AF_INET, type: SOCK_STREAM)
+        let connected = try clientSock.connect(to: serverChannel.localAddress!)
+        XCTAssertEqual(connected, true)
+
+        let clientChannelFuture = try clientSock.withUnsafeFileDescriptor() {
+            ClientBootstrap(group: group).withConnectedSocket(descriptor: $0)
+        }
+        let clientChannel = try clientChannelFuture.wait()
+        _ = clientSock.reapFileDescriptor()
```

Comment:
Rather than doing this, let's call `dup` on the FD that you pass to `withConnectedSocket`, so it's `withConnectedSocket(descriptor: dup($0))`. That avoids the double-closure without adding a method to `BaseSocket`.